### ISSUE TITLE
No validation of NestedAttributes length

### ISSUE THREAD

**Issue by ewbankkit** [hashicorp/terraform#29560]
Relates hashicorp/terraform-plugin-framework#17.
With the latest (fad6afe33058c5a2581d1e424a50c7fc9a842ff3) code it seems that the `MinItems` and `MaxItems` values specified in an `Attribute`'s _NestedAttributesOptions_ (e.g. `ListNestedAttributesOptions`) are not being validated during `terraform plan`.

For example, I can declare an `Attribute` so:

```go
"source_access_configurations": {
	Description: "A list of SourceAccessConfiguration.",
	Attributes: tfsdk.ListNestedAttributes(
		map[string]tfsdk.Attribute{
			"type": {
				Description: "The type of source access configuration.",
				Type:        types.StringType,
				Optional:    true,
			},
			"uri": {
				Description: "The URI for the source access configuration resource.",
				Type:        types.StringType,
				Optional:    true,
			},
		},
		tfsdk.ListNestedAttributesOptions{
			MinItems: 1,
			MaxItems: 2,
		},
	),
	Optional:   true,
},
```

but specify 3 list elements in configuration:

```terraform
  source_access_configurations = [
    {
      type = "VPC_SECURITY_GROUP"
      uri  = aws_security_group.test.id
    },
    {
      type = "VPC_SUBNET"
      uri  = "subnet:${aws_subnet.test.0.id}"
    },
    {
      type = "VPC_SUBNET"
      uri  = "subnet:${aws_subnet.test.1.id}"
    },
  ]
```

and `terraform plan` does not complain:

```
      + source_access_configurations = [
          + {
            + type = "VPC_SECURITY_GROUP"
            + uri  = (known after apply)

            + type = "VPC_SUBNET"
            + uri  = (known after apply)

            + type = "VPC_SUBNET"
            + uri  = (known after apply)
          },
        ]
```

**Comment by jbardin** [comment]
It seems these fields were added to the protocol probably only because they were copied from the Block implementation. The validation against the schema is done during decoding, which happens inside the `hcl` package, but `hcl` itself has no notion of the new nested attribute type schema. The `MinItems`/`MaxItems` fields are a bit awkward on attributes, because attributes are not necessarily containers which can have a number of items to validate (even within Terraform, `NestingSingle` is a bit odd in this way), so this might not make as much sense to pass along to `hcl`. 

So now we have to choose between 3 possible places to handle the new validation:
 - update `hcl` to allow min/max on attributes, even though they may not be container types
 - update Terraform to handle a new form of validation which it doesn't currently have at all
 - update the SDK to handle validation that was designed to be passed on to Terraform

**Comment by apparentlymart** [comment]
`NestingSingle` does indeed have a rather bizarre interpretation of `MinItems` and `MaxItems`. For nested blocks it functions as a funny way to write `Optional: true` or `Optional: false` -- `MinItems: 0` or `MinItems: 1` respectively -- but we don't really need that for attributes of structural types because they can already have `Optional` set anyway.

Given this, my leaning in the ideal case would be to remove these from the protocol entirely and leave this up to the SDK or provider logic to deal with, as we do for most other validation rules that go beyond type checking. Removing them from the protobuf file would be a breaking change in the generated Go stubs though, and so we might prefer to compromise by keeping them around and having Terraform Core be responsible for checking them, as opposed to keeping them around but having Terraform Core _ignore_ them (currently the case).

I would suggest against making `hcldec` aware of this concept, because `hcldec` has no awareness of this Terraform-specific idea of a "normal attribute" vs. a "structual-typed attribute" and it doesn't currently really get into the business of directly validating attribute values aside from type-checking them.

**Comment by paddycarver** [comment]
> Given this, my leaning in the ideal case would be to remove these from the protocol entirely and leave this up to the SDK or provider logic to deal with, as we do for most other validation rules that go beyond type checking. Removing them from the protobuf file would be a breaking change in the generated Go stubs though, and so we might prefer to compromise by keeping them around and having Terraform Core be responsible for checking them, as opposed to keeping them around but having Terraform Core ignore them (currently the case).

We could, as a middle ground, also mark them as deprecated in the protobuf file, remove them from plugin-go, and stop surfacing them in the framework, which would make it a bit more obvious where the framework expects this functionality to be handled (its built-in validation).

They'd technically still be in the protobuf file, but from the officially-supported plugin (as in, supported by the providers org at HashiCorp) point of view, they would no longer exist.

**Comment by ewbankkit** [comment]
The `MaxItems` and `MinItems` fields for nested attributes are used in automatic provider documentation generation:

https://github.com/hashicorp/terraform-plugin-docs/blob/b5dfcab843b2d52b9be4cc49ed006c026ade5ee9/schemamd/write_nested_attribute_type_description.go#L76-L81

However, this is probably a minor consideration compared to the confusion their presence can cause.

**Comment by paddycarver** [comment]
Yeah, the `Description` part of the [`tfsdk.AttributeValidator`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#AttributeValidator) may be able to step in for that, though, and I don't think terraform-plugin-docs would even need a new release, because the schema state field would still be there, it would just always be unset.

**Comment by jbardin** [comment]
Explicitly marking those fields as unused and leaving validation to the provider may be the cleanest way forward here. There is no natural entry point for this in Terraform, because HCL validation is done by `hcldec` and the rest of validation is done by the provider itself. We could insert a new routine before each `ValidateResourceConfig` call, but that feels odd given that it would only be checking Min/Max for nested object containers.

**Comment by jbardin** [comment]
If we want to go ahead and deprecate these 2 fields, I can incorporate those comments into whatever protocol6.1 changes come out of #29572.

**Comment by paddycarver** [comment]
I'd say let's do it. I'll get an issue open on the framework to track adding a validator.
